### DISCUSSION TITLE
Does pgtype support  anyrange @> anyelement operator?

### DISCUSSION THREAD

**Discussion by JesperWe** [jackc/pgx#1864] [General]
(Ref https://www.postgresql.org/docs/15/functions-range.html)

The @> operator can take two forms:
```
anyrange @> anyrange
```
or 
```
anyrange @> anyelement
```
I am currently working with tstzrange, and things are working fine for the first case, `tstzrange @> tstzrange`.
But when trying the second form, `tstzrange @> timestamptz` (meaning "is the timestamp inside the time range") I get the error
```
failed to encode args[0]: unable to encode pgtype.Timestamptz{Time:time.Date(2024, time.January, 4, 15, 38, 35, 189200000, time.Local), InfinityModifier:0, Valid:true} into binary format for tstzrange (OID 3910): unable to encode time.Date(2024, time.January, 4, 15, 38, 35, 189200000, time.Local) into binary format for tstzrange (OID 3910): cannot find encode plan
```

It seems to expect both sides of the operator to be same type?

**Comment by jackc** [comment]
I'm guessing you are doing something like this:

```sql
select time_range @> $1 from reservations
```

pgx uses the extended protocol which parses the statement before binding the values. PostgreSQL returns the expected types of the bind variables. It knows the type of `time_range` since it comes from the table. But it has no way of knowing whether $1 is a `tstzrange` or a `timestamptz`.

Try this instead to let it know the expected type:

```sql
select time_range @> $1::timestamptz from reservations
```

**Comment by JesperWe** [comment]
Yes, that works. Good!
